---
title: 'CA1064：异常应为公共 (代码分析) '
description: 了解代码分析规则 CA1064：异常应该是公共的
ms.date: 11/04/2016
ms.topic: reference
f1_keywords:
- CA1064
- ExceptionsShouldBePublic
helpviewer_keywords:
- ExceptionsShouldBePublic
- CA1064
author: gewarren
ms.author: gewarren
ms.openlocfilehash: 306085feffd6b61b75c268dd1719d037a65a0a8a
ms.sourcegitcommit: 4df8e005c074ceb1f978f007b222fe253be2baf3
ms.translationtype: MT
ms.contentlocale: zh-CN
ms.lasthandoff: 02/04/2021
ms.locfileid: "99546577"
---
# <a name="ca1064-exceptions-should-be-public"></a><span data-ttu-id="78a52-103">CA1064:异常应该是公共的</span><span class="sxs-lookup"><span data-stu-id="78a52-103">CA1064: Exceptions should be public</span></span>

| | <span data-ttu-id="78a52-104">值</span><span class="sxs-lookup"><span data-stu-id="78a52-104">Value</span></span> |
|-|-|
| <span data-ttu-id="78a52-105">**规则 ID**</span><span class="sxs-lookup"><span data-stu-id="78a52-105">**Rule ID**</span></span> |<span data-ttu-id="78a52-106">CA1064</span><span class="sxs-lookup"><span data-stu-id="78a52-106">CA1064</span></span>|
| <span data-ttu-id="78a52-107">**类别**</span><span class="sxs-lookup"><span data-stu-id="78a52-107">**Category**</span></span> |[<span data-ttu-id="78a52-108">Microsoft. Design</span><span class="sxs-lookup"><span data-stu-id="78a52-108">Microsoft.Design</span></span>](design-warnings.md)|
| <span data-ttu-id="78a52-109">**修复是中断或非中断**</span><span class="sxs-lookup"><span data-stu-id="78a52-109">**Fix is breaking or non-breaking**</span></span> |<span data-ttu-id="78a52-110">不间断</span><span class="sxs-lookup"><span data-stu-id="78a52-110">Non-breaking</span></span>|

## <a name="cause"></a><span data-ttu-id="78a52-111">原因</span><span class="sxs-lookup"><span data-stu-id="78a52-111">Cause</span></span>

<span data-ttu-id="78a52-112">非公共异常直接派生自 <xref:System.Exception> 、 <xref:System.SystemException> 或 <xref:System.ApplicationException> 。</span><span class="sxs-lookup"><span data-stu-id="78a52-112">A non-public exception derives directly from <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>.</span></span>

## <a name="rule-description"></a><span data-ttu-id="78a52-113">规则说明</span><span class="sxs-lookup"><span data-stu-id="78a52-113">Rule description</span></span>

<span data-ttu-id="78a52-114">内部异常仅在其自己的内部范围内可见。</span><span class="sxs-lookup"><span data-stu-id="78a52-114">An internal exception is only visible inside its own internal scope.</span></span> <span data-ttu-id="78a52-115">当异常超出内部范围后，只能使用基异常来捕获该异常。</span><span class="sxs-lookup"><span data-stu-id="78a52-115">After the exception falls outside the internal scope, only the base exception can be used to catch the exception.</span></span> <span data-ttu-id="78a52-116">如果内部异常继承自 <xref:System.Exception> 、或， <xref:System.SystemException> <xref:System.ApplicationException> 外部代码将不会有足够的信息来了解如何处理该异常。</span><span class="sxs-lookup"><span data-stu-id="78a52-116">If the internal exception is inherited from <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>, the external code will not have sufficient information to know what to do with the exception.</span></span>

<span data-ttu-id="78a52-117">但是，如果代码有一个公共异常，稍后将其用作内部异常的基，则假设代码更进一步的操作将能够使用基本异常进行智能化。</span><span class="sxs-lookup"><span data-stu-id="78a52-117">But, if the code has a public exception that later is used as the base for a internal exception, it is reasonable to assume the code further out will be able to do something intelligent with the base exception.</span></span> <span data-ttu-id="78a52-118">公共异常将提供比、或提供的信息更多的信息 <xref:System.Exception> <xref:System.SystemException> <xref:System.ApplicationException> 。</span><span class="sxs-lookup"><span data-stu-id="78a52-118">The public exception will have more information than what is provided by <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>.</span></span>

## <a name="how-to-fix-violations"></a><span data-ttu-id="78a52-119">如何解决冲突</span><span class="sxs-lookup"><span data-stu-id="78a52-119">How to fix violations</span></span>

<span data-ttu-id="78a52-120">使异常成为公共异常，或从非、或的公共异常派生内部异常 <xref:System.Exception> <xref:System.SystemException> <xref:System.ApplicationException> 。</span><span class="sxs-lookup"><span data-stu-id="78a52-120">Make the exception public, or derive the internal exception from a public exception that is not <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>.</span></span>

## <a name="when-to-suppress-warnings"></a><span data-ttu-id="78a52-121">何时禁止显示警告</span><span class="sxs-lookup"><span data-stu-id="78a52-121">When to suppress warnings</span></span>

<span data-ttu-id="78a52-122">如果你确定在其自己的内部范围内将捕获私有异常，则禁止显示此规则的消息。</span><span class="sxs-lookup"><span data-stu-id="78a52-122">Suppress a message from this rule if you are sure in all cases that the private exception will be caught within its own internal scope.</span></span>

## <a name="example"></a><span data-ttu-id="78a52-123">示例</span><span class="sxs-lookup"><span data-stu-id="78a52-123">Example</span></span>

<span data-ttu-id="78a52-124">此规则在第一个示例方法 FirstCustomException 上激发，因为 exception 类直接从 Exception 派生，并且是 internal。</span><span class="sxs-lookup"><span data-stu-id="78a52-124">This rule fires on the first example method, FirstCustomException because the exception class derives directly from Exception and is internal.</span></span> <span data-ttu-id="78a52-125">此规则不会在 SecondCustomException 类上激发，因为尽管类也直接派生自异常，但该类声明为公共类。</span><span class="sxs-lookup"><span data-stu-id="78a52-125">The rule does not fire on the SecondCustomException class because although the class also derives directly from Exception, the class is declared public.</span></span> <span data-ttu-id="78a52-126">第三个类还不会激发规则，因为它不是直接从 <xref:System.Exception?displayProperty=fullName> 、 <xref:System.SystemException?displayProperty=fullName> 或派生的 <xref:System.ApplicationException?displayProperty=fullName> 。</span><span class="sxs-lookup"><span data-stu-id="78a52-126">The third class also does not fire the rule because it does not derive directly from <xref:System.Exception?displayProperty=fullName>, <xref:System.SystemException?displayProperty=fullName>, or <xref:System.ApplicationException?displayProperty=fullName>.</span></span>

:::code language="csharp" source="snippets/csharp/all-rules/ca1064.cs" id="snippet1":::
